src/services/refactors/convertStringOrTemplateLiteral.ts

```diff
@@ -0,0 +1,190 @@
+    }
+
+    function treeToArray(node: Node): { nodes: ReadonlyArray<Expression>, containsString: boolean, areOperatorsValid: boolean} {
+        if (isBinaryExpression(node)) {
+            const { nodes: leftNodes, containsString: leftHasString, areOperatorsValid: leftOperatorValid } = treeToArray(node.left);
+            const { nodes: rightNodes, containsString: rightHasString, areOperatorsValid: rightOperatorValid } = treeToArray(node.right);
+
+            if (!leftHasString && !rightHasString) {
+                return { nodes: [node], containsString: false, areOperatorsValid: true };
+            }
+
+            const currentOperatorValid = node.operatorToken.kind === SyntaxKind.PlusToken;
+            const areOperatorsValid = leftOperatorValid && currentOperatorValid && rightOperatorValid;
+
+            return { nodes: leftNodes.concat(rightNodes), containsString: true, areOperatorsValid };
+        }
+
+        return { nodes: [node as Expression], containsString: isStringLiteral(node), areOperatorsValid: true };
+    }
+
+    function concatConsecutiveString(index: number, nodes: ReadonlyArray<Expression>): [number, string] {
+        let text = "";
+
+        while (index < nodes.length && isStringLiteral(nodes[index])) {
+            text = text + decodeRawString(nodes[index].getText());
+            index++;
+        }
+
+        text = escapeText(text);
+        return [index, text];
+    }
+
+    function nodesToTemplate(nodes: ReadonlyArray<Expression>) {
+        const templateSpans: TemplateSpan[] = [];
+        const [begin, headText] = concatConsecutiveString(0, nodes);
+        const templateHead = createTemplateHead(headText);
+
+        if (begin === nodes.length) return createNoSubstitutionTemplateLiteral(headText);
+
+        for (let i = begin; i < nodes.length; i++) {
+            const expression = isParenthesizedExpression(nodes[i]) ? (nodes[i] as ParenthesizedExpression).expression : nodes[i];
+            const [newIndex, subsequentText] = concatConsecutiveString(i + 1, nodes);
+            i = newIndex - 1;
+
+            const templatePart = i === nodes.length - 1 ? createTemplateTail(subsequentText) : createTemplateMiddle(subsequentText);
+            templateSpans.push(createTemplateSpan(expression, templatePart));
+        }
+
+        return createTemplateExpression(templateHead, templateSpans);
+    }
+
+    const octalToUnicode = (_match: string, grp: string) => String.fromCharCode(parseInt(grp, 8));
+
+    function decodeRawString(content: string) {
+        const outerQuotes = /["']((.|\s)*)["']/;
+        const octalEscape = /\\((?:[1-7][0-7]{0,2}|[0-7]{2,3}))/g;
+
+        return content.replace(outerQuotes, (_match, grp) => grp)
+                      .replace(octalEscape, octalToUnicode);
+
+    }
+
+    function escapeText(content: string) {
+        return content.replace("`", "\`")       // back-tick
+                      .replace("${", "$\\{");  // placeholder alike beginning

```

Comment:
i think there's an extra backslash here. maybe it should be `.replace("${", "$\{")`?

Reply:
Unfortunately if I pass as `$\{`, the backslash will not be preserved.
because the parser optimize this `\{` to `{`

Reply:
i tested this locally and this is what happens:
![image](https://user-images.githubusercontent.com/19519347/55096260-ca2fbd00-5076-11e9-8b73-61031e4b4376.png)
so it does seem like there's an extra backslash on the resulting template literal.

Reply:
I understand that `"$\\{"` changes the external behavior

to explain my dilemma
```ts
// before 
const escape = "with ${dollar}"

// after using .replace("${", "$\{")
const escape = `with ${dollar}`

// after using .replace("${", "\${")
const escape = `with ${dollar}`

// after using .replace("${", "$\\{")
const escape = `with $\\{dollar}`

// after using .replace("${", "\\${")
const escape = `with \\${dollar}`
```
using `$\{` or `\${`
* the backslash completely disappears after refactoring
* and therefore, it leaves the template literal in an invalid state

using `$\\{` or `\\${`
* there is one backslash too much 
* however, the template literal is still in valid state

Both approaches change the external behavior.
As it seems, neither `$` nor `{` can be escaped with single backslash.
However, the backtick `` ` `` can be escaped with just a single backslash.

What do you think?
How should I deal with this matter?

Reply:
Since the external behavior cannot be preserved with one or two backslash, I've decided to remove escaping of opening placeholder. `.replace("${", "$\\{")`